integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -386,3 +409,205 @@ def setUp(self):
+      atol: a float as defined in np.allclose. Will be overridden in any
+        comparison between calls to the trace with a `atol` kwarg.
+    """
+    # Trace the test function for each backend.
+    ref_trace = tf_utils.TracedModule(self._ref_module, trace_function)
+    tar_traces = [
+        tf_utils.TracedModule(module, trace_function)
+        for module in self._tar_modules
+    ]
+
+    # Compare each target trace with the reference trace.
+    failed_backend_indices = []
+    for i, tar_trace in enumerate(tar_traces):
+      logging.info("Comparing the reference backend '%s' with '%s'",
+                   ref_trace.backend, tar_trace.backend)
+      traces_match = self._compare_traces(ref_trace, tar_trace, rtol, atol)
+      if not traces_match:
+        failed_backend_indices.append(i)
+
+    # Save the results to disk before validating.
+    trace_dir = os.path.join(self._artifacts_dir, "traces")
+    if not os.path.exists(trace_dir):
+      os.makedirs(trace_dir)
+    ref_trace.save_plaintext(trace_dir, FLAGS.summarize)
+    for tar_trace in tar_traces:
+      tar_trace.save_plaintext(trace_dir, FLAGS.summarize)
+
+    # Validate results.
+    if len(failed_backend_indices) > 0:
+      # Extract info for logging.
+      failed_traces = [tar_traces[b] for b in failed_backend_indices]
+      failed_backends = [trace.backend for trace in failed_traces]
+      failure_info = (
+          "Comparision between the reference backend and the following targets "
+          f"failed: {failed_backends}. The  errors above show the outputs of "
+          "the non-matching calls.")
+
+      # This condition is always True, but is useful for context in the logs.
+      self.assertFalse(len(failed_backends) > 0, failure_info)
+
+  def _compare_traces(self, ref_trace, tar_trace, default_rtol, default_atol):
+    traces_match = True
+    for ref_call, tar_call in zip(ref_trace, tar_trace):
+      ref_call.record_tolerances_if_unset(default_rtol, default_atol)
+      tar_call.record_tolerances_if_unset(default_rtol, default_atol)
+      # Get the potentially overridden tolerances for comparisons of this call.
+      rtol, atol = ref_call.get_tolerances()
+
+      logging.info("Comparing calls to '%s'", ref_call.method)
+      calls_match = self._check_same(ref_call.outputs, tar_call.outputs, rtol,
+                                     atol)
+
+      # Log the inputs and outputs if they don't match.
+      if not calls_match:
+        logging.error("Comparision between '%s' and '%s' failed on method '%s'",
+                      ref_trace.backend, tar_trace.backend, ref_call.method)
+        logging.error("Reference result '%s':\n%s", ref_trace.backend, ref_call)
+        logging.error("Target result '%s':\n%s", tar_trace.backend, tar_call)
+
+      traces_match = traces_match and calls_match
+    return traces_match
+
+  def _check_same(self, ref, tar, rtol, atol):
+    """Checks that ref and tar have identical datastructures and values."""
```

Comment:
I feel like the functionality of checking if two traces match should live with the TraceModule class.

Reply:
That should be fine, since we aren't actually using the `TestCase` assertions in our comparisons. I'll move `TracedModule` into `tf_test_utils.py` then.